test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/SequenceMethodsTests.cs

```diff
@@ -0,0 +1,266 @@
+    /// Tests to check sequence methods support.
+    /// </summary>
+    public class SequenceMethodsTests
+    {
+        private readonly TestContext _ctx;
+
+        public SequenceMethodsTests()
+        {
+            _ctx = new TestContext();
+        }
+
+        [Fact]
+        public void Any()
+        {
+            _ctx.InterceptRequestAndAssertUri("/Customers/$count");
+            _ctx.InterceptRequestAndMockResponse("91");
+            Assert.True(_ctx.Customers.Any());
+
+            _ctx.InterceptRequestAndAssertUri("/Customers/$count?$filter=contains(ContactName,'thisdoesntexist')");
+            _ctx.InterceptRequestAndMockResponse("0");
+            Assert.False(_ctx.Customers.Where(c => c.Name.Contains("thisdoesntexist")).Any());
+        }
+
+        [Fact]
+        public void AnyPredicate()
+        {
+            _ctx.InterceptRequestAndAssertUri("/Customers/$count?$filter=contains(ContactName,'ab')");
+            _ctx.InterceptRequestAndMockResponse("6");
+            Assert.True(_ctx.Customers.Any(c => c.Name.Contains("ab")));
+
+            _ctx.InterceptRequestAndAssertUri("/Customers/$count?$filter=contains(ContactName,'thisdoesntexist')");
+            _ctx.InterceptRequestAndMockResponse("0");
+            Assert.False(_ctx.Customers.Any(c => c.Name.Contains("thisdoesntexist")));
+        }
+
+        [Fact]
+        public void CountPredicate()
+        {
+            _ctx.InterceptRequestAndAssertUri("/Customers/$count?$filter=contains(ContactName,'ab')");
+            _ctx.InterceptRequestAndMockResponse("6");
+            int count = _ctx.Customers.Count(c => c.Name.Contains("ab"));
+            Assert.Equal(6, count);
+        }
+
+        [Fact]
+        public void LongCountPredicate()
+        {
+            _ctx.InterceptRequestAndAssertUri("/Customers/$count?$filter=contains(ContactName,'ab')");
+            _ctx.InterceptRequestAndMockResponse("6");
+            long count = _ctx.Customers.LongCount(c => c.Name.Contains("ab"));
+            Assert.Equal(6, count);
+        }
+
+        [Fact]
+        public void FirstPredicate()
+        {
+            _ctx.InterceptRequestAndAssertUri("/Customers?$filter=ContactName ne 'John'&$top=1");
+            _ctx.InterceptRequestAndMockResponseValue("Customers", "[" + TestContext.MockCustomer1 + "]");
+            Customer customer = _ctx.Customers.First(c => c.Name != "John");
+            Assert.Equal("ALFKI", customer.Id);
+            Assert.Equal("Maria Anders", customer.Name);
+            Assert.Equal("Berlin", customer.City);
+        }
+
+        [Fact]
+        public void FirstPredicate_ThrowsException_WhenNoMatchExists()
+        {
+            _ctx.InterceptRequestAndAssertUri("/Customers?$filter=ContactName eq 'thisdoesntexist'&$top=1");
+            _ctx.InterceptRequestAndMockResponseValue("Customers", "[]");
+            Assert.Throws<InvalidOperationException>(() => _ctx.Customers.First(c => c.Name == "thisdoesntexist"));
+        }
```

Comment:
@joecarl You should make your tests easier to follow. The following block of code has not place in the constructor:
```csharp
            _ctx.BuildingRequest += (obj, args) =>
            {
                if (_assertRequestUriPath == "") return;
                string expectedUri = _rootUriStr + _assertRequestUriPath;
                string actualUri = args.RequestUri.ToString();
                Assert.Equal(expectedUri, actualUri);
                _assertRequestUriPath = "";
            };
```
Your approach makes is making the logic become more complex than it needs to be. Consider this:
```suggestion
        public void FirstPredicate_ThrowsException_WhenNoMatchExists()
        {
            _ctx.InterceptRequestAndMockResponseValue("Customers", "[]");
            _ctx.BuildingRequest += (obj, args) =>
            {
                Assert.EndsWith("/Customers?$filter=ContactName eq 'thisdoesntexist'&$top=1", args.RequestUri.OriginalString);
            };

            Assert.Throws<InvalidOperationException>(() => _ctx.Customers.First(c => c.Name == "thisdoesntexist"));
        }
```
Clean and simple - The assertions are within the test and the logic is not complicated

Reply:
Asserting with `EndsWith` is not correct, that could assert bad URIs. The whole URI (or at least the whole path and queryString) should be checked.
Also, the event listener should be removed after assertion. (Unless tests run over different instances).
Fixing your code in order to address this issues will add more boilerplate to every single test. Are you sure it is worth it?

Reply:
@joecarl xUnit instatiates the test class for each test. So, you would not need to write boilerplate per test case. Currently the `TestContext` does not remove the listener either. So, the test code would not have to change significantly if you removed the `TesContext` class.

I do agree with you that testing the entire url is more robust than using `EndsWith`.

Reply:
Thank you very much for your clarification, it really helps me learn. I understand your point of view, and although I believe a separate class can help with better code reuse and scalability in the future, I also understand your perspectives. In my future contributions, I will avoid creating an additional class to encapsulate this kind of functionality.